src/Compilers/CSharp/Portable/Symbols/Source/ExtensionGroupingInfo.cs

```diff
@@ -0,0 +1,504 @@
+            {
+                return this;
+            }
+
+            IEnumerable<TypeReferenceWithAttributes> ITypeDefinition.Interfaces(EmitContext context)
+            {
+                return SpecializedCollections.EmptyEnumerable<TypeReferenceWithAttributes>();
+            }
+
+            public sealed override bool Equals(object? obj)
+            {
+                // It is not supported to rely on default equality of these Cci objects, an explicit way to compare and hash them should be used.
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            public sealed override int GetHashCode()
+            {
+                // It is not supported to rely on default equality of these Cci objects, an explicit way to compare and hash them should be used.
+                throw ExceptionUtilities.Unreachable();
+            }
+        }
+
+        private sealed class ExtensionGroupingType : ExtensionGroupingOrMarkerType, IComparable<ExtensionGroupingType>
+        {
+            private readonly string _name;
+            public readonly ImmutableArray<ExtensionMarkerType> ExtensionMarkerTypes;
+            private readonly ImmutableArray<ExtensionGroupingTypeParameter> _typeParameters;
+
+            public ExtensionGroupingType(string name, MultiDictionary<string, SourceNamedTypeSymbol> extensionMarkerTypes)
+            {
+                _name = name;
+
+                var builder = ArrayBuilder<ExtensionMarkerType>.GetInstance(extensionMarkerTypes.Count);
+
+                foreach (var pair in extensionMarkerTypes)
+                {
+                    builder.Add(new ExtensionMarkerType(this, pair.Key, pair.Value));
+                }
+
+                builder.Sort();
+                ExtensionMarkerTypes = builder.ToImmutableAndFree();
+
+                _typeParameters = ExtensionMarkerTypes[0].UnderlyingExtensions[0].Arity != 0 ?
+                    ((INestedTypeDefinition)ExtensionMarkerTypes[0].UnderlyingExtensions[0].GetCciAdapter()).GenericParameters.SelectAsArray(static (p, @this) => new ExtensionGroupingTypeParameter(@this, p), this) :
+                    [];
+            }
+
+            int IComparable<ExtensionGroupingType>.CompareTo(ExtensionGroupingType? other)
+            {
+                Debug.Assert(other is { });
+                return ExtensionMarkerTypes[0].CompareTo(other.ExtensionMarkerTypes[0]);
+            }
+
+            protected override IEnumerable<IGenericTypeParameter> GenericParameters => _typeParameters;
+
+            protected override ushort GenericParameterCount => (ushort)ExtensionMarkerTypes[0].UnderlyingExtensions[0].Arity;
+
+            protected override bool IsAbstract => false;
+
+            protected override bool IsSealed => true;
+
+            protected override ITypeDefinition ContainingTypeDefinition => ExtensionMarkerTypes[0].UnderlyingExtensions[0].ContainingType!.GetCciAdapter();
```

Comment:
>! [](http://example.com/codeflow?start=137&length=1)

This should be an assertion (here or possibly in the constructor)

Reply:
> This should be an assertion (here or possibly in the constructor)

I do not think its worth the ceremony. The fact is obvious by construction, i.e. these are nested types symbols.